File: src/lexer.ts
enum TokenType {
  NUM = 0, // 数字 0-9.
  PLUS, // 加号 +
  MINUS, // 减号 -
  MULT, // 乘号 *
  DIV, // 除号 /
  MOD, // 求模符号 %
  EQ, // 等于号 =
  GT, // 大于号 >
  LT, // 小于号 <
  AND, // 与 &
  OR, // 或 |
  XOR, // 异或 ^
  NOT, // 取反 ~
  NE, // 不等于
  EXTENDS,
  LPARENT , // 左圆括号 (
  RPARENT , // 右圆括号 )
  LBRACES, // 左花括号 {
  RBRACES , // 右花括号 }
  LBRACKETS, // 左方括号 [
  RBRACKETS, // 右方括号 ]
  KEYWORD_IF, // 关键字 if
  KEYWORD_ELSE, // 关键字 else
  KEYWORD_WHILE, // 关键字 while
  KEYWORD_RETURN, // 关键字 return
  KEYWORD_CALL, // 关键字 call
  KEYWORD_FOR, // 关键字 for
  KEYWORD_CHANNEL, // 关键字 channel,
  KEYWORD_VARIABLE,
  SYMBOL,
  RANGE,
  DOT,
  EOF
}

class Token {
  constructor(public readonly token: TokenType, public readonly value: string|number) {}
}

export function Lexer(code: string) {
  // 数字
  let NUMERIC = "0123456789";
  // 十六进制
  let HEX = "abcdefABCDEF";
  // 字母
  let ALPHNUM = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz$";
  // 操作符
  let OPERATE = "+-*/%~<!|~=>@.";
  // 空白字符
  let WHITESPACE = "\t\n ";
  let tokens: Token[] = [];
  let codeIndex = 0;
  while (codeIndex < code.length) {
    let singleChar = code[codeIndex];
    if (NUMERIC.includes(singleChar)) {
      let n = codeIndex + 1;
      while (n < code.length && (NUMERIC + "_." + HEX).includes(code[n])) {
        n++;
      }
      if(code.slice(codeIndex, n).includes(".")){
          
          tokens.push(new Token(TokenType.NUM, parseFloat(code.slice(codeIndex, n))));
        }else{
          tokens.push(new Token(TokenType.NUM, parseInt(code.slice(codeIndex, n))));
      }
      codeIndex = n;
    } else if (WHITESPACE.includes(singleChar)) {
      codeIndex++;
    } else if (OPERATE.includes(singleChar)) {
      switch (singleChar) {
        case "&":
          tokens.push(new Token(TokenType.AND, singleChar));
          codeIndex++;
          break;
        case "|":
          tokens.push(new Token(TokenType.OR, singleChar));
          codeIndex++;
          break;
        case "^":
          tokens.push(new Token(TokenType.XOR, singleChar));
          codeIndex++;
          break;
        case "%":
          tokens.push(new Token(TokenType.MOD, singleChar));
          codeIndex++;
          break;
        case "~":
          tokens.push(new Token(TokenType.NOT, singleChar));
          codeIndex++;
          break;
        case "=":
          tokens.push(new Token(TokenType.PLUS, singleChar));
          codeIndex++;
          break;
        case ".":
            if(code.slice(codeIndex,codeIndex + 3)=== "..."){
                tokens.push(new Token(TokenType.RANGE,code.slice(codeIndex,codeIndex+3)))
                codeIndex+=3;
            }else{
                codeIndex ++;
            }
        case "@":
          if (code.slice(codeIndex, codeIndex + 2) === "@>") {
            tokens.push(new Token(TokenType.EXTENDS, code.slice(codeIndex,codeIndex+2)));
            codeIndex += 2;
          }else{
            codeIndex++;
          }
          break;
        case "+":
          if (code.slice(codeIndex, codeIndex + 2) === "+=") {
            tokens.push(new Token(TokenType.PLUS, code.slice(codeIndex, codeIndex + 2)));
            codeIndex += 2;
          } else {
            tokens.push(new Token(TokenType.PLUS, singleChar));
            codeIndex++;
          }
          break;
        case "-":
          if (code.slice(codeIndex, codeIndex + 2) === "-=") {
            tokens.push(new Token(TokenType.MINUS, code.slice(codeIndex, codeIndex + 2)));
            codeIndex += 2;
          } else {
            tokens.push(new Token(TokenType.MINUS, singleChar));
            codeIndex++;
          }
          break;
        case "/":
          if (code.slice(codeIndex, codeIndex + 2) === "/=") {
            tokens.push(new Token(TokenType.DIV, code.slice(codeIndex, codeIndex + 2)));
            codeIndex += 2;
          } else {
            tokens.push(new Token(TokenType.DIV, singleChar));
            codeIndex++;
          }
          break;
        case "*":
          if (code.slice(codeIndex, codeIndex + 2) === "*=") {
            tokens.push(new Token(TokenType.MULT, code.slice(codeIndex,codeIndex+2)));
            codeIndex += 2;
          } else {
            tokens.push(new Token(TokenType.MULT, singleChar));
            codeIndex++;
          }
          break;
        case "%":
          if (code.slice(codeIndex, codeIndex + 2) === "%=") {
            tokens.push(new Token(TokenType.MOD, code.slice(codeIndex, codeIndex + 2)));
            codeIndex += 2;
          } else {
            tokens.push(new Token(TokenType.MOD, singleChar));
            codeIndex ++;
          }
          break;
        case ">":
          if (code.slice(codeIndex, codeIndex + 2) === ">=") {
            tokens.push(new Token(TokenType.GT, code.slice(codeIndex, codeIndex + 2)));
            codeIndex += 2;
          } else {
            tokens.push(new Token(TokenType.GT, singleChar));
            codeIndex++;
          }
          break;
        case "<":
          if (code.slice(codeIndex, codeIndex + 2) === "<=") {
            tokens.push(new Token(TokenType.LT, code.slice(codeIndex, codeIndex + 2)));
            codeIndex += 2;
          } else {
            tokens.push(new Token(TokenType.LT, singleChar));
            codeIndex++;
          }
          break;
        case "!":
          if (code.slice(codeIndex, codeIndex + 2) === "!=") {
            tokens.push(new Token(TokenType.NE, code.slice(codeIndex, codeIndex + 2)));
            codeIndex += 2;
          } else {
            tokens.push(new Token(TokenType.NOT, singleChar));
            codeIndex++;
          }
          break;
      }
    } else if ("{}".includes(singleChar) || "[]".includes(singleChar) || "()".includes(singleChar)) {
      switch (singleChar) {
        case "(":
          tokens.push(new Token(TokenType.LPARENT, singleChar));
          break;
        case ")":
          tokens.push(new Token(TokenType.RPARENT, singleChar));
          break;
        case "{":
          tokens.push(new Token(TokenType.LBRACES, singleChar));
          break;
        case "}":
          tokens.push(new Token(TokenType.RBRACES, singleChar));
          break;
        case "[":
          tokens.push(new Token(TokenType.LBRACKETS, singleChar));
          break;
        case "]":
          tokens.push(new Token(TokenType.RBRACKETS, singleChar));
          break;
      }
      codeIndex++;
    } else if (ALPHNUM.includes(singleChar)) {
      let w = codeIndex + 1;
      while (w < code.length && (ALPHNUM+NUMERIC+ "_").includes(code[w])) {
        w++;
      }
      let word = code.slice(codeIndex, w);
      switch (word) {
        case "IF":
        case "if":
          tokens.push(new Token(TokenType.KEYWORD_IF, word));
          break;
        case "ELSE":
        case "else":
          tokens.push(new Token(TokenType.KEYWORD_ELSE, word));
          break;
        case "WHILE":
        case "while":
          tokens.push(new Token(TokenType.KEYWORD_WHILE, word));
          break;
        case "for":
        case "FOR":
          tokens.push(new Token(TokenType.KEYWORD_FOR, word));
          break;
        case "CALL":
        case "call":
          tokens.push(new Token(TokenType.KEYWORD_CALL, word));
          break;
        case "CHANNEL":
        case "channel":
          tokens.push(new Token(TokenType.KEYWORD_CHANNEL, word));
          break;
        case "var":
        case "VAR":
          tokens.push(new Token(TokenType.KEYWORD_VARIABLE, word));
          break;
        default:
          tokens.push(new Token(TokenType.SYMBOL, word));
      }
      codeIndex = w;
    }else if(";".includes(singleChar)){
        tokens.push(new Token(TokenType.EOF,singleChar))
        codeIndex++;
    } else {
      throw new Error("未知的字符: " + singleChar);
    }
  }
  return tokens;
}


let l = Lexer(`
    ((12+33.5+12)+12/34*32)+66.796
    if 1233 else 32
    channel A @> B 12 != 3334;
    12>= 23;
    4>= 3;
    34<= 23;
    sim += 23;
    for i ... 20{
        12 JK 33;
        c3 d5 s6;
        $0 c4 Margin;
        let s = 33;
    }
`);
// this will be return:
// [
//   Token { token: 14, value: '(' },
//   Token { token: 14, value: '(' },
//   Token { token: 0, value: '12' },
//   Token { token: 1, value: '+' },
//   Token { token: 0, value: '33.5' },
//   Token { token: 1, value: '+' },
//   Token { token: 0, value: '12' },
//   Token { token: 15, value: ')' },
//   Token { token: 1, value: '+' },
//   Token { token: 0, value: '12' },
//   Token { token: 4, value: '/' },
//   Token { token: 0, value: '34' },
//   Token { token: 3, value: '*' },
//   Token { token: 0, value: '32' },
//   Token { token: 15, value: ')' },
//   Token { token: 1, value: '+' },
//   Token { token: 0, value: '66.796' },
//   Token { token: 20, value: 'if' },
//   Token { token: 0, value: '12' },
//   Token { token: 7, value: '>' },
//   Token { token: 0, value: '33' },
//   Token { token: 21, value: 'else' },
//   Token { token: 0, value: '32' },
//   Token { token: 26, value: 'channel' },
//   Token { token: 0, value: '3' }
// ]

console.log(l);

File: src/lexer.py
# 词法分析器的作用是将代码分割成一个个可以解释的token,
# 用以供后续的Parser进行解析

from enum import Enum
from typing import List

class TokenType(Enum):
    NUM = 0 # 数字
    PLUS = 1 # 加号
    MINUS = 2 # 减号
    MULT = 3  # 乘号
    DIV = 4 # 除号
    MOD = 5 # 求模符号
    EQ = 6 # 等于号
    GT = 7 # 大于号
    LT = 8 # 小于号
    AND = 9 # 与
    OR = 10 # 或
    XOR = 11 # 异或
    NOT = 13 # 取反
    LPARENT = 14
    RPARENT = 15
    LBRACES = 16 # 
    RBRACES = 17
    LBRACKETS = 18
    RBRACKETS = 19

class Token:
    token_type:TokenType
    value:str
    def __init__(self,token_type:TokenType,value:str) -> None:
        self.token_type = token_type
        self.value = value
    def __repr__(self) -> str:
        return f"Toekn({self.token_type}-> { self.value })"

DIGITS = "0123456789"
HEX = "0123456789abcdefABCDEF"
OPERATORS = "=+-*/%><^|&~"
BRACES = "{}"
BRACKETS = "[]"
PARENT = "()"
WHITESPACE = "\t\n "

# 词法分析
def lexer(code):
    token = []
    i = 0
    while i < len(code):
        single_char = code[i]
        if single_char in DIGITS:
            # 如果是数字,则将连续的数字合并成一个token
            j = i + 1
            DIGITS_ = DIGITS+"."
            while(j<len(code) and code[j] in DIGITS_): 
                # 如果词中含有数字和点,都归类为数字
                j += 1 # 记录含有数字的下标 方便后续的取出子字符串
            token.append(Token(TokenType.NUM,code[i:j]))
            i = j # 重置全局的下标计数
        elif(single_char in HEX):
            sub_index = i + 1
            while(sub_index < len(code) and code[sub_index] in HEX):
                sub_index += 1
            token.append(Token(TokenType.NUM,code[i:sub_index]))
        elif(single_char in OPERATORS):
            # 如果是运算符 那就每个运算符单独作为一个token
            if(single_char == "+"):
                token.append(Token(TokenType.PLUS,single_char))
            elif(single_char == "-"):
                token.append(Token(TokenType.MINUS,single_char))
            elif(single_char == "*"):
                token.append(Token(TokenType.MULT,single_char))
            elif(single_char == "/"):
                token.append(Token(TokenType.DIV,single_char))
            elif(single_char == "%"):
                token.append(Token(TokenType.MOD,single_char))
            elif(single_char == ">"):
                token.append(Token(TokenType.GT,single_char))
            elif(single_char == "<"):
                token.append(Token(TokenType.LT,single_char))
            elif(single_char == "~"):
                token.append(Token(TokenType.NOT,single_char))
            elif(single_char == "&"):
                token.append(Token(TokenType.AND,single_char))
            elif(single_char == "|"):
                token.append(Token(TokenType.OR,single_char))
            elif(single_char == "^"):
                token.append(Token(TokenType.XOR,single_char))
            i+=1
        elif(single_char in PARENT or single_char in BRACES or single_char in BRACKETS):
            if(single_char == "("):
                token.append(Token(TokenType.LPARENT,single_char))
            elif(single_char == ")"):
                token.append(Token(TokenType.RPARENT,single_char))
            elif(single_char == "{"):
                token.append(Token(TokenType.LBRACES,single_char))
            elif(single_char == "}"):
                token.append(Token(TokenType.RBRACES,single_char))
            elif(single_char == "["):
                token.append(Token(TokenType.LBRACKETS,single_char))
            elif(single_char== "]"):
                token.append(Token(TokenType.RBRACKETS,single_char))
            i += 1
        elif(single_char in WHITESPACE):
            # 如果是空白字符继续向字符串后面找
            i += 1
        else:
            raise Exception(f"Invalid character: {single_char}")
    return token

# 定义语法分析函数
def parse(tokens:List[Token]):
    # 定义一个指针，指向当前正在处理的 Token
    i = 0

    # 定义一个辅助函数，用于读取下一个 Token
    def next_token():
        nonlocal i
        if i < len(tokens):
            i += 1
            return tokens[i - 1]
        else:
            return None

    # 定义表达式语法规则的解析函数
    def expr():
        left = term()
        token = next_token()
        while token and token.type in (TokenType.PLUS, TokenType.MINUS):
            op = token.type
            right = term()
            left = (op, left, right)
            token = next_token()
        return left

    # 定义项语法规则的解析函数
    def term():
        left = factor()
        token = next_token()
        while token and token.type in (TokenType.MULT, TokenType.DIV, TokenType.MOD):
            op = token.type
            right = factor()
            left = (op, left, right)
            token = next_token()
        return left

    # 定义因子语法规则的解析函数
    def factor():
        token = next_token()
        if token.type == TokenType.NUM:
            return float(token.value)
        elif token.type == TokenType.LPAREN:
            result = expr()
            if next_token().type != TokenType.RPAREN:
                raise Exception("Expected closing parenthesis")
            return result
        else:
            raise Exception("Expected number or opening parenthesis")


if __name__ == '__main__':
    tokens = lexer("""
    1+2/4 = 23.3412
    12 > 45
    62 < 43
    (23+45)
    ^12
    ~333
    """)
    parser_tokens = parse(tokens)
    print(tokens)
    